src/Controller/Action/ApplyCouponAction.php

```diff
@@ -54,27 +69,67 @@ public function __invoke(Request $request): Response
             $form->handleRequest($request);
 
             if ($form->isSubmitted() && $form->isValid()) {
-                /** @var OrderInterface $cart */
-                $cart = $this->cartContext->getCart();
-                Assert::isInstanceOf($cart, OrderInterface::class);
+                Assert::notNull($command->coupon);
+
+                if (null !== $this->promotionCouponEligibilityChecker && !$this->promotionCouponEligibilityChecker->isEligible($cart, $command->coupon)) {
+                    $this->addFlash($request, 'error', 'setono_sylius_coupon_url_application.coupon_not_eligible');
+
+                    return $this->redirect($command);
+                }
 
                 $cart->setPromotionCoupon($command->coupon);
 
+                $totalBefore = $cart->getTotal();
                 $this->orderProcessor->process($cart);
+                $totalAfter = $cart->getTotal();
 
                 $this->getManager($cart)->flush();
 
-                $session = $request->getSession();
-                if ($session instanceof Session) {
-                    $session->getFlashBag()->add('success', 'setono_sylius_coupon_url_application.coupon_applied');
+                $totalDiff = abs($totalBefore - $totalAfter);
+
+                $request->getSession()->set('sscua_coupon_applied', true);
```

Comment:
I'm not so crazy about setting a session variable here. Atleast, maybe we can remove it, when we know we've shown the flash message?

Reply:
Me neither. I hate it and wanted to use URL parameters, but because of the way the form works (it works with GET) I couldn't do that.